Foundation/NSJSONSerialization.swift

```diff
@@ -511,15 +515,33 @@ private struct JSONWriter {
             } else {
                 writer(",")
             }
-            
-            if key is String {
-                try serializeString(key as! String)
+
+            if let key = key as? String {
+                try serializeString(key)
             } else {
                 throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "NSDictionary key must be NSString"])
             }
             pretty ? writer(": ") : writer(":")
             try serializeJSON(value)
         }
+
+        if sortedKeys {
+            let elems = try dict.sorted(by: { a, b in
+                guard let a = a.key as? String,
+                    let b = b.key as? String else {
+                        throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "NSDictionary key must be NSString"])
+                }
+                return a < b
```

Comment:
I don't think this will apply the exact same sorting strategy as we do on Darwin platforms. We are currently doing the equivalent of:

```objc
NSString * const key1 = ...
NSString * const key2 = ...
const NSRange range = NSMakeRange(0, key1.length);
NSLocale * const locale = [NSLocale systemLocale];
return [key1 compare:key2 options:NSNumericSearch | NSCaseInsensitiveSearch | NSForcedOrderingSearch range:range locale:locale];
```

Especially for archiving purposes, we should try to be consistent here.